### ISSUE TITLE
Freyd: method installed for IsInjective matches more than one declaration

### ISSUE THREAD

**Issue by sebastianpos** [homalg-project/CAP_project#586]
During my investigation of this problem, I produced the (so far) following minimal example:
```
LoadPackage( "GradedRingForHomalg" );
LoadPackage( "FinS" );
CategoryOfSkeletalFinSets( : overhead := false );
#I  method installed for IsInjective matches more than one declaration
```

If I omit the overhead option, then the info will not be printed. This hints at a problem of the overhead option.

**Comment by sebastianpos** [comment]
This bug is very strange.

The new minimal example is:
```
LoadPackage( "homalg" );
LoadPackage( "FinS" );
CategoryOfSkeletalFinSets( : overhead := false );
```
and I suspected a conflict with `IsInjective` for `IsHomalgObject`.

The warning is triggered when `FinSets` calls `AddIsInjective` which in turn triggers in `CAP`
https://github.com/homalg-project/CAP_project/blob/6c95daf14d869d56589758313f8e088bb0162c00/CAP/gap/InstallAdds.gi#L516
the installation of `IsInjective` with the filter
`[ <Category "(SkeletalFinSetsObjectFilter and IsCapCategoryObject)"> ]`.

`IsHomalgObject` does not conflict with the filter above, at least calling `IsHomalgObject` on the initial object of the output category yields `false`. Nevertheless, the warning occurs at this point. So maybe I'm not aware of the correct conflict here.

If I omit the overhead option,  `FinSets` calls `AddIsInjective` which in turn triggers in `CAP`
https://github.com/homalg-project/CAP_project/blob/6c95daf14d869d56589758313f8e088bb0162c00/CAP/gap/InstallAdds.gi#L438
the installation of `IsInjective`. The filter at that point is the same as above.
If I manually interrupt at that point and do the installation manually with this filter, no warning is triggered.

I'm puzzled.

**Comment by zickgraf** [comment]
I think the difference is that `install_method` in line 438 is `InstallMethodWithCache`, whereas line 516 simply uses `InstallMethod`.

**Comment by zickgraf** [comment]
Further findings: The issue seems to be related to line: https://github.com/homalg-project/CAP_project/blob/6c95daf14d869d56589758313f8e088bb0162c00/CAP/gap/InstallAdds.gi#L430

If I exclude this line for `IsProjective`, the error does not happen anymore.

**Comment by sebastianpos** [comment]
I used `InstallMethod` instead of `InstallMethodWithCache` in my manual experiment. No warning was triggered.

Unfortunately, I cannot reproduce the effect that you saw of 

https://github.com/homalg-project/CAP_project/blob/6c95daf14d869d56589758313f8e088bb0162c00/CAP/gap/InstallAdds.gi#L430

I put an `Assert( 0, install_name <> "IsInjective" )` before that line and it is never triggered. (I assume you meant `IsInjective` instead of `IsProjective`).

**Comment by zickgraf** [comment]
No, I indeed meant `IsProjective`. Here is a new minimal example:
```
LoadPackage( "homalg" );
LoadPackage( "CAP" );
cat := CreateCapCategory();
InstallTrueMethod( IsProjective, ObjectFilter(cat) and IsCapCategoryObject );
InstallMethod( IsInjective, [ObjectFilter(cat) and IsCapCategoryObject ], ReturnFalse );
```

**Comment by zickgraf** [comment]
Even more minimal example:
```
LoadPackage( "ToolsForHomalg" );

DeclareCategory( "IsHomalgObjectOrMorphism",
        IsExtLElement and
        IsStructureObjectOrObjectOrMorphism );

DeclareCategory( "IsHomalgObject",
        IsHomalgObjectOrMorphism and
        IsStructureObjectOrObject and
        IsAdditiveElementWithZero );

DeclareProperty( "IsProjective",
        IsHomalgObject );
DeclareProperty( "IsInjective",
        IsHomalgObject );

LoadPackage( "CAP" );
cat := CreateCapCategory();

InstallTrueMethod( IsProjective, ObjectFilter(cat) and IsCapCategoryObject );
InstallMethod( IsInjective, [ObjectFilter(cat) and IsCapCategoryObject ], ReturnFalse );
```
I think that's the point where I will stop for now, as I have the feeling this goes all the way down into GAP itself.

**Comment by sebastianpos** [comment]
Thanks for your investigation. I cut it down to the following example, that does not have anything to do with `CAP`, `IsProjective`, or `IsInjective`:
```

DeclareCategory( "IsExampleObjA", IsObject );

DeclareCategory( "IsExampleObjB", IsObject );

DeclareProperty( "IsPropX",
        IsExampleObjA );
DeclareProperty( "IsPropY",
        IsExampleObjA );

DeclareProperty( "IsPropX",
        IsExampleObjB );
DeclareProperty( "IsPropY",
        IsExampleObjB );

InstallTrueMethod( IsPropX, IsExampleObjB );

InstallMethod( IsPropY, [ IsExampleObjB ], ReturnFalse );
```

**Comment by mohamed-barakat** [comment]
Thank you very much for investigating. I am happy it has nothing to do with homalg :) It looks like a small bug in the logic triggering the warning in GAP.

**Comment by mohamed-barakat** [comment]
Maybe @fingolfin could have a look before creating an issue on GAP.

**Comment by fingolfin** [comment]
I didn't check this in detail (not at a computer right now) but it sounds as if you rediscovered https://github.com/gap-system/gap/issues/1649

**Comment by zickgraf** [comment]
Thanks for the pointer :-) I can confirm that when calling gap with the `-N` switch to disable hidden implications, the issue does not occur.

**Comment by zickgraf** [comment]
I think after reading the linked issue, I now understand what's going on:
```
DeclareProperty( "IsPropX",
        IsExampleObjA );
```
creates a (hidden) implication `IsPropX => IsExampleObjA`.
```
InstallTrueMethod( IsPropX, IsExampleObjB );
```
creates a (real) implication `IsExampleObjB => IsPropX`.

Thus, we get `IsExampleObjB => IsExampleObjA`, which is why
```
InstallMethod( IsPropY, [ IsExampleObjB ], ReturnFalse );
```
matches both declarations of `IsPropY`. If we turn off hidden implications, this does not happen.

Running the Freyd tests with `gap -N` gives the following output:
```
InstallMethod warning: Wedge at /homalg_project/Modules/gap/ExteriorAlgebra.gi:392 required filter IsExteriorPowerElement for 1st argument does not match any declaration
InstallMethod warning: SingleValueOfExteriorPowerElement at /homalg_project/Modules/gap/ExteriorAlgebra.gi:420 required filter IsExteriorPowerElement for 1st argument does not match any declaration
InstallMethod warning: ExteriorPowerElementDual at /homalg_project/Modules/gap/ExteriorAlgebra.gi:449 required filter IsExteriorPowerElement for 1st argument does not match any declaration
InstallMethod warning: TransposedMatImmutable at /QPA2/lib/matrix.gi:215 required filter IsIdentityMatrix for 1st argument does not match any declaration
InstallMethod warning: TransposedMatImmutable at /QPA2/lib/matrix.gi:223 required filter IsZeroMatrix for 1st argument does not match any declaration
InstallMethod warning: MatElm at /QPA2/lib/matrix.gi:257 required filter IsIdentityMatrix for 1st argument does not match any declaration
InstallMethod warning: MatElm at /QPA2/lib/matrix.gi:263 required filter IsZeroMatrix for 1st argument does not match any declaration
InstallMethod warning: OppositeQuiver at /QPA2/lib/quiver.gi:1530 required filter IsProductQuiver for 1st argument does not match any declaration
InstallMethod warning: PreImagesRepresentative at /QPA2/lib/algebra.gi:1686 required filter IsNaturalHomomorphismByIdeal for 1st argument does not match any declaration
InstallMethod warning: IsHereditaryAlgebra at /QPA2/lib/special-algebras.gi:766 required filter IsAdmissibleQuiverAlgebra for 1st argument does not match any declaration
```

It seems like we have to fix these warnings anyway if the removal of hidden implications continues, so we could fix these and run GAP with the `-N` switch by default in the tests.

@fingolfin: Is the removal of hidden implications advanced enough that is makes sense to run GAP with the `-N` switch in our CI and report issues upstream if we encounter warnings?

**Comment by zickgraf** [comment]
This is now fixed as a byproduct of 8fbeab367f452b1885125361791ff621c23c898b.
